README.md
# Football Manager Data Science (FMDS)
<img align="center" width="500" height="200" src="./media/logo.png">

FMDS is a Python script designed for Football Manager data analysis and reporting. This script processes exported data from Football Manager, calculates player attributes based on predefined weights, and generates interactive HTML reports using DataTables integration.

First thing first, Thanks to [squirrel_plays](https://www.youtube.com/@squirrel_plays_fof4318) for the Inspiration and their hard work on the football manager views, Their code is under `./original.py`.

## Prerequisites
- Windows 11 (May work on 10 & 7)

## Install
- [Download the latest zip from Github](https://github.com/ktasper/FMDS/releases)
- Extract the zip to a new folder, I do `Documents/FMDS`
- Run the `exe`
- Or `python main.py` if doing it via the source code.

## Usage
1. Export Data from Football Manager
- Export squad data from Football Manager and save it to a folder.
    - I use the `all_attribute_squad` for my squad view.
      - The views can be found in the `fm_files` folder
    - Select all players with shift click
    - `CTRL + P` > Web Page
    - Save this file where-ever you want, I have a folder on my Desktop called `Documents/FMDS-Data` so its easy to do in the FM UI.
        - Don't forget to give it a name you understand.

3. Define Position Weights (Optional)
- Customize player attribute weights for different positions by editing the `positions.json` file. Follow the specified format to define scoring categories such as "essential," "core," and "secondary."
    - You can add as many or as little categories as you want.
    - [The position short code and what they mean can be found here](https://github.com/ktasper/FMDS/wiki/Attribute-Meanings)
    


We also create a few custom metrics:
```
"Spd" = "Pac" + "Acc" / 2
"Work" ="Wor" + "Sta" / 2
"SetP" ="Jum" + "Bra" / 2
```


4. Run the Script
- Execute the fm_data_science.py script.
- Run the `fmds.exe`
    - or `python main.py`

<img align="center" width="600" height="600" src="./media/ui-screenshot.webp">
The program will ask you for the following:

- `FM Export Path` - This is where you saved the exported views (IE player / Team stats) via the views under `fm_files`
- `Output Dir` - This is where the generated reports go.
- `Save / Load Config` - Lets you save or load the above settings so you do need to set them each time (NOTE: you need to press load to load the settings.)

5. Generated HTML Report
- The script generates an HTML report (output.html) with DataTables integration. Open this file in a web browser to view the interactive report. The file gets saved to the `generated_data` folder



## Customization

### Modifying Position Weights

Customize player attribute weights for different positions by editing the positions.json file. Update the attributes and weights based on your game preferences.

[The position short code and what they mean can be found here](https://github.com/ktasper/FMDS/wiki/Attribute-Meanings)

<img align="center" width="800" height="500" src="./media/postions_json.png">

### Custom Generated Views

Supported names are:
```
[
    "Reg", "Inf", "Name", "Age", "Wage", "Transfer Value", "Nat", "2nd Nat",
    "Position", "Personality", "Media Handling", "Av Rat", "Left Foot", "Right Foot",
    "Height", "1v1", "Acc", "Aer", "Agg", "Agi", "Ant", "Bal", "Bra", "Cmd", "Cnt",
    "Cmp", "Cro", "Dec", "Det", "Dri", "Fin", "Fir", "Fla", "Han", "Hea", "Jum", "Kic",
    "Ldr", "Lon", "Mar", "OtB", "Pac", "Pas", "Pos", "Ref", "Sta", "Str", "Tck", "Tea",
    "Tec", "Thr", "TRO", "Vis", "Wor", "UID", "Cor", "Club"
]
```

## Notes
The script only works on the latest file in the `fm_export_path`.




Changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Released]
## [0.1.0] - 11-11-23

### Added
- This CHANGELOG file to hopefully serve as an evolving example of a
  standardized open source project CHANGELOG.
- Added FMDS program
- Added build step to generate zips for users who do not want to run `python`
## [Unreleased]

original.py
#!/usr/bin/env python
# coding: utf-8
import pandas as pd
import glob
import os

# finds most recent file in specified folder
list_of_files = glob.glob(
    os.path.join("/Users/archie/Desktop/FM files on desktop", "*")
)
latest_file = max(list_of_files, key=os.path.getctime)
print(latest_file)

# Read HTML file exported by FM - in this case an example of an output from the squad page
# This reads as a list, not a dataframe
squad_rawdata_list = pd.read_html(
    latest_file, header=0, encoding="utf-8", keep_default_na=False
)

# turn the list into a dataframe
squad_rawdata = squad_rawdata_list[0]

# Calculate simple speed and workrate scores
squad_rawdata["Spd"] = (squad_rawdata["Pac"] + squad_rawdata["Acc"]) / 2
squad_rawdata["Work"] = (squad_rawdata["Wor"] + squad_rawdata["Sta"]) / 2
squad_rawdata["SetP"] = (squad_rawdata["Jum"] + squad_rawdata["Bra"]) / 2

# calculates gk score
squad_rawdata["gk_essential"] = (squad_rawdata["Agi"] + squad_rawdata["Ref"]) * 5
squad_rawdata["gk_core"] = (
    squad_rawdata["1v1"]
    + squad_rawdata["Ant"]
    + squad_rawdata["Cmd"]
    + squad_rawdata["Cnt"]
    + squad_rawdata["Kic"]
    + squad_rawdata["Pos"]
) * 3
squad_rawdata["gk_secondary"] = (
    squad_rawdata["Acc"]
    + squad_rawdata["Aer"]
    + squad_rawdata["Cmp"]
    + squad_rawdata["Dec"]
    + squad_rawdata["Fir"]
    + squad_rawdata["Han"]
    + squad_rawdata["Pas"]
    + squad_rawdata["Thr"]
    + squad_rawdata["Vis"]
) * 1
squad_rawdata["gk"] = (
    (squad_rawdata["gk_essential"])
    + (squad_rawdata["gk_core"])
    + (squad_rawdata["gk_secondary"])
) / 37
squad_rawdata.gk = squad_rawdata.gk.round(1)
# for others: squad_rawdata['gk_core'] = ( squad_rawdata[''] + squad_rawdata[''] + squad_rawdata['']+ squad_rawdata['']+ squad_rawdata['']+ squad_rawdata['']+ squad_rawdata['']) / 2


# calculates fb score
squad_rawdata["fb_essential"] = (
    squad_rawdata["Wor"]
    + squad_rawdata["Acc"]
    + squad_rawdata["Pac"]
    + squad_rawdata["Sta"]
)
squad_rawdata["fb_core"] = (
    squad_rawdata["Cro"]
    + squad_rawdata["Dri"]
    + squad_rawdata["Mar"]
    + squad_rawdata["OtB"]
    + squad_rawdata["Tck"]
    + squad_rawdata["Tea"]
)
squad_rawdata["fb_secondary"] = (
    squad_rawdata["Agi"]
    + squad_rawdata["Ant"]
    + squad_rawdata["Cnt"]
    + squad_rawdata["Dec"]
    + squad_rawdata["Fir"]
    + squad_rawdata["Pas"]
    + squad_rawdata["Pos"]
    + squad_rawdata["Tec"]
)
squad_rawdata["fb"] = (
    (squad_rawdata["fb_essential"] * 5)
    + (squad_rawdata["fb_core"] * 3)
    + (squad_rawdata["fb_secondary"] * 1)
) / 46
squad_rawdata.fb = squad_rawdata.fb.round(1)


# calculates cb score
squad_rawdata["cb_core"] = (
    squad_rawdata["Cmp"]
    + squad_rawdata["Hea"]
    + squad_rawdata["Jum"]
    + squad_rawdata["Mar"]
    + squad_rawdata["Pas"]
    + squad_rawdata["Pos"]
    + squad_rawdata["Str"]
    + squad_rawdata["Tck"]
    + squad_rawdata["Pac"]
) / 9
squad_rawdata["cb_secondary"] = (
    squad_rawdata["Agg"]
    + squad_rawdata["Ant"]
    + squad_rawdata["Bra"]
    + squad_rawdata["Cnt"]
    + squad_rawdata["Dec"]
    + squad_rawdata["Fir"]
    + squad_rawdata["Tec"]
    + squad_rawdata["Vis"]
) / 8
squad_rawdata["cb"] = (squad_rawdata["cb_core"] * 0.75) + (
    squad_rawdata["cb_secondary"] * 0.25
)
squad_rawdata.cb = squad_rawdata.cb.round(1)


# calculates dm score
squad_rawdata["dm"] = (
    (squad_rawdata["Wor"] * 5)
    + (squad_rawdata["Pac"] * 5)
    + (squad_rawdata["Sta"] * 3)
    + (squad_rawdata["Pas"] * 3)
    + (squad_rawdata["Tck"] * 2)
    + (squad_rawdata["Ant"] * 2)
    + (squad_rawdata["Cnt"] * 2)
    + (squad_rawdata["Pos"] * 2)
    + (squad_rawdata["Bal"] * 2)
    + (squad_rawdata["Agi"] * 2)
    + (squad_rawdata["Tea"] * 1)
    + (squad_rawdata["Fir"] * 1)
    + (squad_rawdata["Mar"] * 1)
    + (squad_rawdata["Agg"] * 1)
    + (squad_rawdata["Cmp"] * 1)
    + (squad_rawdata["Dec"] * 1)
    + (squad_rawdata["Str"] * 1)
) / 35
squad_rawdata.dm = squad_rawdata.dm.round(1)

# calculates segundo volante on attack score
squad_rawdata["vol"] = (
    (squad_rawdata["Wor"] * 5)
    + (squad_rawdata["Pac"] * 5)
    + (squad_rawdata["Sta"] * 3)
    + (squad_rawdata["Pas"] * 3)
    + (squad_rawdata["Tck"] * 2)
    + (squad_rawdata["Ant"] * 2)
    + (squad_rawdata["Cnt"] * 2)
    + (squad_rawdata["Pos"] * 2)
    + (squad_rawdata["Tea"] * 2)
    + (squad_rawdata["Fir"] * 1)
    + (squad_rawdata["Mar"] * 1)
    + (squad_rawdata["Agg"] * 1)
    + (squad_rawdata["Cmp"] * 1)
    + (squad_rawdata["Dec"] * 1)
    + (squad_rawdata["Str"] * 1)
) / 32
squad_rawdata.vol = squad_rawdata.vol.round(1)

# calculates box2box score
squad_rawdata["box2"] = (
    (squad_rawdata["Pas"] * 5)
    + (squad_rawdata["Wor"] * 5)
    + (squad_rawdata["Sta"] * 4)
    + (squad_rawdata["Tck"] * 3)
    + (squad_rawdata["OtB"] * 3)
    + (squad_rawdata["Tea"] * 3)
    + (squad_rawdata["Vis"] * 2)
    + (squad_rawdata["Str"] * 2)
    + (squad_rawdata["Dec"] * 2)
    + (squad_rawdata["Pos"] * 2)
    + (squad_rawdata["Pac"] * 2)
    + (squad_rawdata["Agg"] * 1)
    + (squad_rawdata["Ant"] * 1)
    + (squad_rawdata["Fin"] * 1)
    + (squad_rawdata["Lon"] * 1)
    + (squad_rawdata["Cmp"] * 1)
    + (squad_rawdata["Acc"] * 1)
    + (squad_rawdata["Bal"] * 1)
    + (squad_rawdata["Fir"] * 1)
    + (squad_rawdata["Dri"] * 1)
    + (squad_rawdata["Tec"] * 1)
)
squad_rawdata.box2 = squad_rawdata.box2.round(0)

# calculates winger score
squad_rawdata["w_core"] = (
    squad_rawdata["Acc"]
    + squad_rawdata["Cro"]
    + squad_rawdata["Dri"]
    + squad_rawdata["OtB"]
    + squad_rawdata["Pac"]
    + squad_rawdata["Tec"]
) / 6
squad_rawdata["w_secondary"] = (
    squad_rawdata["Agi"]
    + squad_rawdata["Fir"]
    + squad_rawdata["Pas"]
    + squad_rawdata["Sta"]
    + squad_rawdata["Wor"]
) / 5
squad_rawdata["w"] = (squad_rawdata["w_core"] * 0.75) + (
    squad_rawdata["w_secondary"] * 0.25
)
squad_rawdata.w = squad_rawdata.w.round(1)

# calculates inverted winger score
squad_rawdata["amrl"] = (
    (squad_rawdata["Acc"] * 5)
    + (squad_rawdata["Pac"] * 5)
    + (squad_rawdata["Wor"] * 5)
    + (squad_rawdata["Dri"] * 3)
    + (squad_rawdata["Pas"] * 3)
    + (squad_rawdata["Tec"] * 3)
    + (squad_rawdata["OtB"] * 3)
    + (squad_rawdata["Cro"] * 1)
    + (squad_rawdata["Fir"] * 1)
    + (squad_rawdata["Cmp"] * 1)
    + (squad_rawdata["Dec"] * 1)
    + (squad_rawdata["Vis"] * 1)
    + (squad_rawdata["Agi"] * 1)
    + (squad_rawdata["Sta"] * 1)
) / 34
squad_rawdata.amrl = squad_rawdata.amrl.round(1)

# calculates amc score
squad_rawdata["amc"] = (
    (squad_rawdata["Vis"] * 4)
    + (squad_rawdata["OtB"] * 4)
    + (squad_rawdata["Pas"] * 4)
    + (squad_rawdata["Dec"] * 3)
    + (squad_rawdata["Ant"] * 3)
    + (squad_rawdata["Cmp"] * 3)
    + (squad_rawdata["Tec"] * 3)
    + (squad_rawdata["Dri"] * 1)
    + (squad_rawdata["Fir"] * 1)
    + (squad_rawdata["Fla"] * 1)
    + (squad_rawdata["Lon"] * 1)
    + (squad_rawdata["Agi"] * 1)
    + (squad_rawdata["Fin"] * 1)
)
squad_rawdata.amc = squad_rawdata.amc.round(0)


# calculates striker score
squad_rawdata["str_core"] = (
    squad_rawdata["Cmp"]
    + squad_rawdata["Fin"]
    + squad_rawdata["OtB"]
    + squad_rawdata["Pac"]
) / 4
squad_rawdata["str_secondary"] = (
    squad_rawdata["Acc"]
    + squad_rawdata["Agi"]
    + squad_rawdata["Ant"]
    + squad_rawdata["Bal"]
    + squad_rawdata["Dec"]
    + squad_rawdata["Dri"]
    + squad_rawdata["Fir"]
    + squad_rawdata["Pas"]
    + squad_rawdata["Sta"]
    + squad_rawdata["Tec"]
    + squad_rawdata["Wor"]
) / 11
squad_rawdata["str"] = (squad_rawdata["str_core"] * 0.5) + (
    squad_rawdata["str_secondary"] * 0.5
)
squad_rawdata.str = squad_rawdata.str.round(1)

# builds squad dataframe using only columns that will be exported to HTML
squad = squad_rawdata[
    [
        "Inf",
        "Name",
        "Age",
        "Club",
        "Transfer Value",
        "Wage",
        "Nat",
        "Position",
        "Personality",
        "Media Handling",
        "Left Foot",
        "Right Foot",
        "Spd",
        "Jum",
        "Str",
        "Work",
        "Height",
        "gk",
        "fb",
        "cb",
        "vol",
        "str",
    ]
]

# taken from here: https://www.thepythoncode.com/article/convert-pandas-dataframe-to-html-table-python
# creates a function to make a sortable html export


def generate_html(dataframe: pd.DataFrame):
    # get the table HTML from the dataframe
    table_html = dataframe.to_html(table_id="table", index=False)
    # construct the complete HTML with jQuery Data tables
    # You can disable paging or enable y scrolling on lines 20 and 21 respectively
    html = f"""
    <html>
    <header>
        <link href="https://cdn.datatables.net/1.11.5/css/jquery.dataTables.min.css" rel="stylesheet">
    </header>
    <body>
    {table_html}
    <script src="https://code.jquery.com/jquery-3.6.0.slim.min.js" integrity="sha256-u7e5khyithlIdTpu22PHhENmPcRdFiHRjhAuHcs05RI=" crossorigin="anonymous"></script>
    <script type="text/javascript" src="https://cdn.datatables.net/1.11.5/js/jquery.dataTables.min.js"></script>
    <script>
        $(document).ready( function () {{
            $('#table').DataTable({{
                paging: false,
                order: [[12, 'desc']],
                // scrollY: 400,
            }});
        }});
    </script>
    </body>
    </html>
    """
    # return the html
    return html


# generates random file name for write-out of html file
import uuid

filename = str(uuid.uuid4()) + ".html"
filename

# creates a sortable html export from the dataframe 'squad'
html = generate_html(squad)
open(filename, "w", encoding="utf-8").write(html)
macos_filename = "/Users/millie/Desktop/FM files on desktop" + filename


build_tools/build.ps1
remove-item package/ -Force
pyinstaller --onefile --windowed --icon=media/icon.ico main.py
copy-item media/ dist/ -Recurse -Force
copy-item positions.json dist -Force
copy-item views.json dist -Force
move-item dist/main.exe dist/FMDS.exe -Force
new-item -Path "./package/FMDS" -ItemType "directory" -Force
Get-Item -Path dist\* | Move-Item -Destination package\FMDS\ -Force
remove-item dist/ -Force
new-item -Path "./package/FMDS/fm_files/filters/" -ItemType "directory" -Force
new-item -Path "./package/FMDS/fm_files/views/" -ItemType "directory" -Force
Get-Item -Path fm_files\filters\* | Copy-Item -Destination package\FMDS\fm_files\filters\ -Force
Get-Item -Path fm_files\views\* | Copy-Item -Destination package\FMDS\fm_files\views\ -Force

Compress-Archive -Path package/FMDS -DestinationPath package/FMDS.zip

main.py
import glob
import json
import os
import uuid
import sys
import pandas as pd
import tkinter as tk
from tkinter import filedialog
from tkinter import messagebox
from tkinter import ttk


fm_export_path = ""
class CustomLogger:
    def __init__(self, text_widget):
        self.text_widget = text_widget

    def write(self, message):
        self.text_widget.insert(tk.END, message)
        self.text_widget.see(tk.END)  # Autoscroll to the bottom
        self.text_widget.update_idletasks()  # Update the Tkinter GUI

    def flush(self):
        pass

class ToolTip:
    def __init__(self, widget, text):
        self.widget = widget
        self.text = text
        self.tooltip = None
        self.widget.bind("<Enter>", self.display_tooltip)
        self.widget.bind("<Leave>", self.hide_tooltip)

    def display_tooltip(self, event):
        x, y, _, _ = self.widget.bbox("insert")
        x += self.widget.winfo_rootx() + 25
        y += self.widget.winfo_rooty() + 25
        self.tooltip = tk.Toplevel(self.widget)
        self.tooltip.wm_overrideredirect(True)
        self.tooltip.wm_geometry(f"+{x}+{y}")
        label = tk.Label(
            self.tooltip,
            text=self.text,
            background="#ffffe0",
            relief="solid",
            borderwidth=1,
        )
        label.pack()

    def hide_tooltip(self, event):
        if self.tooltip:
            self.tooltip.destroy()
            self.tooltip = None


def gen_html(dataframe: pd.DataFrame):
    # get the table HTML from the dataframe
    table_html = dataframe.to_html(table_id="table", index=False)
    # construct the complete HTML with jQuery Data tables
    # You can disable paging or enable y scrolling on lines 20 and 21 respectively
    html = f"""
    <html>
    <header>
        <link href="https://cdn.datatables.net/1.11.5/css/jquery.dataTables.min.css" rel="stylesheet">
    </header>
    <body>
    {table_html}
    <script src="https://code.jquery.com/jquery-3.6.0.slim.min.js" integrity="sha256-u7e5khyithlIdTpu22PHhENmPcRdFiHRjhAuHcs05RI=" crossorigin="anonymous"></script>
    <script type="text/javascript" src="https://cdn.datatables.net/1.11.5/js/jquery.dataTables.min.js"></script>
    <script>
        $(document).ready( function () {{
            $('#table').DataTable({{
                paging: false,
                order: [[12, 'desc']],
                // scrollY: 400,
            }});
        }});
    </script>
    </body>
    </html>
    """
    # return the html
    return html


def weight_calc(position, category_id, position_weights: json, squad_rawdata):
    """
    This takes in a position as defined in the json
    And a category_id (IE: essential, core, secondary) and provides a consistent way of calculating a players ability
    based on given attributes and weights from the json
    We do this via the value of the attribute * the weight divided by the number of items * the weight
    """
    try:
        attributes = position_weights["positions"][position][category_id]["attributes"]
    except KeyError:
        attributes = []
    if attributes:
        try:
            weight = position_weights["positions"][position][category_id]["weight"]
        except KeyError:
            weight = 0
        # calculate the normaliser via the number of items * the weight
        normaliser = (
            len(position_weights["positions"][position][category_id]["attributes"])
            * weight
        )
        # print(
        #    f"""
        #      Debug:
        #      Position: {position}
        #      Cat: {category_id}
        #      Debug weight: {weight}
        #      normaliser: {normaliser}"""
        # )
        attribute_values = (
            sum(squad_rawdata[attribute] for attribute in attributes) * weight
        )
        if normaliser != 0:
            v = attribute_values / normaliser
            return v
        else:
            return 0
    else:
        # print(
        #    f"""
        #      Debug:
        #      Position: {position}
        #      Cat: {category_id}
        #      Debug weight: Not set (0)
        #      normaliser: Not Set (0)"""
        # )
        return 0


def main(fm_export_path, output_path):
    # finds most recent file in specified folder
    export_dir: str = fm_export_path

    list_of_files = glob.glob(os.path.join(export_dir, "*"))
    try:
        latest_file = max(list_of_files, key=os.path.getctime)
        latest_filename = os.path.basename(latest_file).replace(".html", "")
    except ValueError:
        print(f"No files found in {export_dir}")
        latest_filename = None

    # print(f"Debug Latest File: {latest_filename}")

    # Read HTML file exported by FM - in this case an example of an output from the squad page
    # This reads as a list, not a dataframe
    squad_rawdata_list: list = pd.read_html(
        latest_file, header=0, encoding="utf-8", keep_default_na=False
    )

    # transform the list into a dataframe
    squad_rawdata = squad_rawdata_list[0]

    # Calculate simple speed and workrate scores
    squad_rawdata["Spd"] = (squad_rawdata["Pac"] + squad_rawdata["Acc"]) / 2
    squad_rawdata["Work"] = (squad_rawdata["Wor"] + squad_rawdata["Sta"]) / 2
    squad_rawdata["SetP"] = (squad_rawdata["Jum"] + squad_rawdata["Bra"]) / 2

    # Keep a list of all the headers we want to use in the final view
    final_view_headers = []

    # Add all the exported view names for the final view
    allowed_view_items = [
        "Reg",
        "Inf",
        "Name",
        "Age",
        "Wage",
        "Transfer Value",
        "Nat",
        "2nd Nat",
        "Position",
        "Personality",
        "Media Handling",
        "Av Rat",
        "Left Foot",
        "Right Foot",
        "Height",
        "1v1",
        "Acc",
        "Aer",
        "Agg",
        "Agi",
        "Ant",
        "Bal",
        "Bra",
        "Cmd",
        "Cnt",
        "Cmp",
        "Cro",
        "Dec",
        "Det",
        "Dri",
        "Fin",
        "Fir",
        "Fla",
        "Han",
        "Hea",
        "Jum",
        "Kic",
        "Ldr",
        "Lon",
        "Mar",
        "OtB",
        "Pac",
        "Pas",
        "Pos",
        "Ref",
        "Spd",
        "Work",
        "Sta",
        "Str",
        "Tck",
        "Tea",
        "Tec",
        "Thr",
        "TRO",
        "Vis",
        "Wor",
        "UID",
        "Cor",
        "Club",
    ]

    # Add the views from the "views.json" to the final views list if its in the allowed list
    try:
        with open("views.json", "r") as views_json:
            core_views = json.load(views_json)
    except FileNotFoundError:
        print("Could not find the views.json file")
    for view_item in core_views["order"]:
        if view_item in allowed_view_items:
            final_view_headers.append(view_item)
        else:
            print(f"Ignoring {view_item} since it is not in the allowed views list")

    # Load position weights from JSON file
    try:
        with open("positions.json", "r") as json_file:
            position_weights = json.load(json_file)

            # Retrieve position category from JSON configuration
            position_category = list(
                position_weights["positions"].keys()
            )  # Get the key from the "positions" dictionary
            for position in position_category:
                squad_rawdata[f"{position}"] = 0
                # Now get each scoring category from the json ie: "core, essential etc"
                scoring_categories = list(
                    position_weights["positions"][position].keys()
                )
                for category in scoring_categories:
                    # Add the data back to the data frame after we have run some calculations on it
                    squad_rawdata[f"{position}"] += weight_calc(
                        position, category, position_weights, squad_rawdata
                    )
                squad_rawdata[position] = squad_rawdata[position].round(1)
                final_view_headers.append(position)
    except FileNotFoundError:
        print("Could not find the positions.json file")

    squad = squad_rawdata[final_view_headers]
    # Create the dir if it does not exist
    if not os.path.exists(f"{output_path}/fmds_fmds_generated_data/"):
        os.makedirs(f"{output_path}/fmds_generated_data/")
    # generates random file name for write-out of html file based on the file we use

    filename = str(
        f"{output_path}/fmds_generated_data/{latest_filename}-{uuid.uuid4()}.html"
    )
    # creates a sortable html export from the dataframe 'squad'
    html = gen_html(squad)
    open(filename, "w", encoding="utf-8").write(html)
    print(f"Saved file: {filename}")


# Function to select export path using file dialog
def select_fm_export_path():
    global fm_export_path
    fm_export_path = filedialog.askdirectory()
    fm_export_path_label.config(text="Export Path: " + fm_export_path)


# Function to select output directory using file dialog
def select_output_dir():
    global output_dir
    output_dir = filedialog.askdirectory()
    output_dir_label.config(text="Output Directory: " + output_dir)


# Function to generate the HTML and perform other operations
def generate_html():
    try:
        main(fm_export_path, output_dir)
    except Exception as e:
        messagebox.showerror("Error", f"An error occurred: {str(e)}")


# Function to save configuration settings to JSON file
def save_config_to_json():
    try:
        config_data = {
            "fm_export_path": fm_export_path,
            "output_dir": output_dir
            # Add more configuration parameters here if needed
        }
        with open("config.json", "w") as config_file:
            json.dump(config_data, config_file)
        print("Configuration saved to config.json.")
    except NameError:
        print("Please set the paths before saving!")


# Function to load configuration settings from JSON file
def load_config_from_json():
    global fm_export_path, output_dir  # Declare fm_export_path and output_dir as global variables
    try:
        with open("config.json", "r") as config_file:
            config_data = json.load(config_file)
            fm_export_path = config_data.get("fm_export_path")
            output_dir = config_data.get("output_dir")
            if fm_export_path and output_dir:
                fm_export_path_label.config(text="Export Path: " + fm_export_path)
                output_dir_label.config(text="Output Directory: " + output_dir)
                print("Configuration loaded from config.json.")
            else:
                print("Invalid configuration data in config.json.")
    except FileNotFoundError:
        print("No config.json file found.")
    return fm_export_path, output_dir  # Return the values if needed elsewhere 


root = tk.Tk()
root.iconbitmap("media/icon.ico")
root.title("FMDS")

# Create a Text widget for logging
log_text = tk.Text(root, wrap=tk.WORD, height=20, width=80)
log_text.pack(padx=10, pady=10, expand=True, fill=tk.BOTH)

# Redirect sys.stdout and sys.stderr to the custom logger
custom_logger = CustomLogger(log_text)
sys.stdout = custom_logger
sys.stderr = custom_logger


# Configure styles
style = ttk.Style()
style.configure("TButton", padding=10, font=("Arial", 12))

# Frame for export path
export_frame = ttk.Frame(root, padding=10, relief="solid", borderwidth=1)
export_frame.pack(pady=10, padx=10, fill="x")
fm_export_path_button = ttk.Button(
    export_frame, text="Select FM Exported Data Path", command=select_fm_export_path
)
fm_export_path_button.pack(side="left")
fm_export_path_label = ttk.Label(export_frame, text="Export Path: Not Selected")
fm_export_path_label.pack(side="left")
tooltip_export = ToolTip(fm_export_path_button, "Select the export path containing the exported views from Football Manager")

# Frame for output directory
output_frame = ttk.Frame(root, padding=10, relief="solid", borderwidth=1)
output_frame.pack(pady=10, padx=10, fill="x")
output_dir_button = ttk.Button(
    output_frame, text="Select Output Directory", command=select_output_dir
)
output_dir_button.pack(side="left")
output_dir_label = ttk.Label(output_frame, text="Output Directory: Not Selected")
output_dir_label.pack(side="left")
tooltip_output = ToolTip(output_dir_button, "Select the output directory, This will create a fmds_generated_data folder")

# Frame for save and load config buttons
config_frame = ttk.Frame(root, style="TFrame")
config_frame.pack(pady=10, padx=10, fill="x")

# Button to load config from JSON
load_button = ttk.Button(config_frame, text="Load Config", command=load_config_from_json, style="TButton")
load_button.pack(side="left")

# Button to save config to JSON
save_button = ttk.Button(config_frame, text="Save Config", command=save_config_to_json, style="TButton")
save_button.pack(side="left")

# Action button with an icon
action_button = ttk.Button(
    root, text="Generate", command=generate_html, compound="left", style="TButton"
)
action_button.pack(pady=10)

root.mainloop()